fix: ensure -zlazy for systems that default to -znow

xandris opened the pull request:
I see #1706 and #1722. this is more minimal change for systems that default to znow

xandris committed:
build: Ensure lazy binding for operating systems defaulting to '-z now'

Many operating systems default to '-z now' in their linker flags to
harden binaries using Relocation Read-Only (or RELRO) [1].  Such as,
Fedora [2,3] and Gentoo [4,5,6,7,8,9].

Some, like Fedora, pass it through the LDFLAGS environment variable
that's usually injected into the clang(1) and gcc(1) compiler drivers'
command line by build systems like Meson.  Some, like Gentoo, use a
Clang configuration file or GCC specs instead.

The reference Go toolchain (ie., gc, not gccgo) in its external linking
mode always uses the clang(1) or gcc(1) compiler drivers [10], and
whether '-z now' gets used or not depends on which of the above
mechanisms are in use.

The Meson adapter scripts in Toolbx to connect Meson to 'go build' and
'go run', since Meson doesn't natively support Go [11], don't respect
the LDFLAGS environment variable.  So, anything passed through it gets
filtered out from the linker flags.  However, the default Clang
configuration files or the built-in GCC specs can't be avoided.  So,
if '-z now' is specified there, then it gets used.  This leads to a
broken toolbox(1) binary and fails to generate the shell completions.

[1] https://www.redhat.com/en/blog/hardening-elf-binaries-using-relocation-read-only-relro

[2] Fedora redhat-rpm-config commit 796b80f2f49f2301
    https://src.fedoraproject.org/rpms/redhat-rpm-config/c/796b80f2f49f2301

[3] Fedora redhat-rpm-config commit d9235d2d90873ff6
    https://src.fedoraproject.org/rpms/redhat-rpm-config/c/d9235d2d90873ff6
    https://bugzilla.redhat.com/show_bug.cgi?id=1192183
    https://fedoraproject.org/wiki/Changes/Harden_All_Packages

[4] Gentoo ebuild commit 8bfd8afef6dd8c66
    https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=8bfd8afef6dd8c66
    https://bugs.gentoo.org/876923

[5] Gentoo gcc-patches commit 8ffd428773f8e1e6
    https://github.com/gentoo/gcc-patches/commit/8ffd428773f8e1e6

[6] Gentoo ebuild commit 718448f923ae9302
    https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=718448f923ae9302

[7] Gentoo gcc-patches commit 98c1d96dbf03d704
    https://github.com/gentoo/gcc-patches/commit/98c1d96dbf03d704
    https://bugs.gentoo.org/876923

[8] Gentoo ebuild commit adf44514cf590a86
    https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=adf44514cf590a86
    https://bugs.gentoo.org/876923

[9] Gentoo ebuild commit 975678507aa3cebb
    https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=975678507aa3cebb
    https://bugs.gentoo.org/876923

[10] https://pkg.go.dev/cmd/link

[11] https://github.com/mesonbuild/meson/issues/123

https://github.com/containers/toolbox/issues/1706

Signed-off-by: Alexandra Parker <alex.iris.parker@gmail.com>

debarshiray commented on src/go-build-wrapper:
```diff
@@ -116,7 +116,7 @@ dynamic_linker="/run/host$dynamic_linker_canonical_dirname/$dynamic_linker_basen
         $tags \
         -trimpath \
-        -ldflags "-extldflags '-Wl,-dynamic-linker,$dynamic_linker -Wl,-rpath,/run/host$libc_dir_canonical_dirname -Wl,--export-dynamic -Wl,--unresolved-symbols=ignore-in-object-files' -linkmode external -X github.com/containers/toolbox/pkg/version.currentVersion=$4" \
+        -ldflags "-extldflags '-Wl,-z,lazy,-dynamic-linker,$dynamic_linker,-rpath,/run/host$libc_dir_canonical_dirname,--export-dynamic,--unresolved-symbols=ignore-in-object-files' -linkmode external -X github.com/containers/toolbox/pkg/version.currentVersion=$4" \
```
I suspect we have a typo here that explains the CI failures.

Shouldn't it be `-Wl,-dynamic-linker,$dynamic_linker -Wl,-z,lazy`?

xandris replied:
it's okay to stack up linker arguments in `-Wl`. the commas just separate successive argv elements that ld receives, so it's not necessary to have multiple `-Wl` flags... but it doesn't hurt either

i think the problem was actually a missing comma in my change to `src/meson_generate_completions.py`! that caused Python to concatenate the `lazy` with the `.` that was intended for `go run`. sorry about that

debarshiray replied:
Oops!  Yes, you are right.  I shouldn't do fly-by reviews so late at night.

debarshiray replied:
I took the liberty to split the stacking up of external linker flags into a separate commit.  These build details are hairy and the changes are hard to read, so this will make it easier for our (mainly my) future selves.  :)

debarshiray reviewed (changes requested):
Thanks for looking into this, @xandris , and my apologies for the delay!  I always shudder at this linker hackery.  :)

debarshiray commented:
recheck

debarshiray reviewed (commented):
It looks like this fixes https://github.com/containers/toolbox/issues/1706  Could you please link to the issue instead of the PR in the commit message?

xandris committed:
build: Tweak the external linker flags for brevity

It's not necessary to prefix each external linker flag with its own -Wl
option.  One -Wl option can carry everything because the commas will be
used to split the string.

https://github.com/containers/toolbox/issues/1706

debarshiray commented:
Thanks for updating this PR, @xandris !  Let me try to rebase this against `main`.  I am temporarily marking this as a draft while I do so.

debarshiray commented:
I took the liberty to dump all the references you dug up for me in https://github.com/containers/toolbox/issues/1706 in the commit message for the sake of my future self.

debarshiray committed:
build: Notify distributors that the '-z now' linker flag is unsupported

This should have been part of commit 83f28c52e47c2d44.

https://github.com/containers/toolbox/issues/1706

xandris commented:
:eyes:

debarshiray commented:
> Build failed. https://gateway-cloud-softwarefactory.apps.ocp.cloud.ci.centos.org/zuul/t/local/buildset/2c9465a0661544b89f4c3822984dc1fe
> 
> ❌ [system-test-fedora-rawhide-commands-options ](https://gateway-cloud-softwarefactory.apps.ocp.cloud.ci.centos.org/zuul/t/local/build/a5430751ccf9421b8e0ec44da70f4e63) **TIMED_OUT** in 2h 00m 26s
> ❌ [system-test-fedora-44-commands-options ](https://gateway-cloud-softwarefactory.apps.ocp.cloud.ci.centos.org/zuul/t/local/build/4895bed699e047b5bf406f78340286d4) **TIMED_OUT** in 1h 45m 29s

These are the same CI failures that showed up recently and is being debugged.  See https://github.com/containers/toolbox/pull/1802 and https://github.com/containers/toolbox/pull/1805

So, let's ignore them here.

debarshiray commented:
Thanks for all the insight and investigation, @xandris !
